docs/content/basics/testing_theory.md

```diff
@@ -1,57 +1,163 @@
 
-# Where it all starts
+
+## Where it all starts
 <i>Why do I need testing?</i> Users of your application/system/whatever should receive a quality product. 
 <br/>Thus, testing is about ensuring the quality of the product. Testing also includes creating a test plan, creating/conducting tests themselves and analyzing results of testing.
 
-Let's focus on the topic of creating tests. To begin with, let's see the testing pyramid.
+There are different means to classify tests. We will focus on the following
+
+1. *Granularity*
+2. *Running environment*
+3. *Target*
+
+## Classifying tests by granularity
+To begin with, let's see the testing pyramid.
 
 ![Testing pyramid](../images/pyramid.png "Testing pyramid")
 
+1. **End-to-End tests (E2E)**: these tests verify that all components in every single layer of the application work together as expected. They allow to test User stories - how our users are going to use the application. This makes them incredibly important both for business and for an ordinary developer or tester.
+   However, this is the most unstable type of test, and the cost of creating and maintaining them is high. That's because there are a lot of factors we cannot control (e.g. network failures, clean user state before and after the test, etc.).
+2. **Integration tests**: they focus on validating the interaction of 2 (or more) entities at once at the same time, but not the full system as with E2E tests. Most components involved are not mocks or stubs, but doubles or fakes.
+3. **Unit tests**: they only cover the smallest testable unit, usually a function, or view.
+
 The width of each block is actually the ratio of the number of different types of tests to each other. 
-For example, a lot of Unit tests is usually considered correct in the pyramid, but much less UI tests. 
+For example, a lot of Unit tests is usually considered correct in the pyramid, but much less E2E tests. 
 <br/><br/><b>This is due to two key parameters - stability and the cost of supporting each type of testing.</b>
-<br/>Unit tests have the highest stability, they are the fastest and they have the lowest cost of support, however, using Unit tests you won't be able to check the login flow of your application (which for example contains 5 screens).
-<br/><br/>We will reveal each type of testing in more detail, for this we will go from the bottom up.
+<br/>Unit tests have the highest stability, they are the fastest and they have the lowest cost of support. However, Unit tests do not intend to verify Use stories e.g. the login flow of your application (which for example contains 5 screens).
+<br/><br/>
 
 !!! Disclaimer
 
+    Testing tools and frameworks can be classified regarding their access to the implementation details as follows:
+
     <b>White-box testing</b> is a type of testing in which we have full access to the implementation and can interact with it. We know which output data will be with given input data.
```

Comment:
This could be another section/subsection/different page on its own Basically the tools/frameworks like Espresso, UiAutomator... etc are those categorised into any of these groups

Reply:
but anyway I think it's better to keep that info in first article